iroh-gossip/src/proto/hyparview.rs

```diff
@@ -0,0 +1,707 @@
+        else if message.ttl == self.config.passive_random_walk_length {
+            self.add_passive(message.peer.id, message.peer.data.clone(), io);
+        }
+        // "iii) The time to live field is decremented."
+        // "iv) If, at this point, n has not been inserted
+        // in p’s active view, p will forward the request to a random node in its active view
+        // (different from the one from which the request was received)."
+        if !self.active_view.contains(&message.peer.id) {
+            match self
+                .active_view
+                .pick_random_without(&[&sender], &mut self.rng)
+            {
+                None => {
+                    // TODO: I think this is unreachable!() but will have to check, maybe decrease
+                    // to warn
+                    unreachable!("no peers in active view but also did not add this node on forward join, this should not happen");
+                }
+                Some(next) => {
+                    let message = Message::ForwardJoin(ForwardJoin {
+                        peer: message.peer,
+                        ttl: message.ttl.next(),
+                    });
+                    io.push(OutEvent::SendMessage(*next, message));
+                }
+            }
+        }
+    }
+
+    fn on_neighbor(&mut self, from: PA, details: Neighbor, now: Instant, io: &mut impl IO<PA>) {
+        self.pending_neighbor_requests.remove(&from);
+        // "A node q that receives a high priority neighbor request will always accept the request, even
+        // if it has to drop a random member from its active view (again, the member that is dropped will
+        // receive a Disconnect notification). If a node q receives a low priority Neighbor request, it will
+        // only accept the request if it has a free slot in its active view, otherwise it will refuse the request."
+        match details.priority {
+            Priority::High => {
+                self.add_active(from, details.data, Priority::High, now, io);
+            }
+            Priority::Low if !self.active_is_full() => {
+                self.add_active(from, details.data, Priority::Low, now, io);
+            }
+            _ => {}
+        }
+    }
+
+    fn peer_info(&self, id: &PA) -> Option<PeerInfo<PA>> {
+        self.peer_data.get(id).map(|data| PeerInfo {
+            id: *id,
+            data: data.clone(),
+        })
+    }
+
+    fn insert_peer_info(&mut self, peer_info: PeerInfo<PA>, io: &mut impl IO<PA>) {
+        let old = self.peer_data.remove(&peer_info.id);
+        let same = matches!(old, Some(old) if old == peer_info.data);
+        if !same {
+            io.push(OutEvent::PeerData(peer_info.id, peer_info.data.clone()));
+        }
+        self.peer_data.insert(peer_info.id, peer_info.data);
+    }
+
+    fn on_shuffle(&mut self, from: PA, shuffle: Shuffle<PA>, io: &mut impl IO<PA>) {
```

Comment:
my interpretation is that shuffle should be responded when it cannot be forwarded as well. From the paper:
>A node q that receives a Shuffle request will first decrease its time to live. If the time to live of the message is greater than zero **and the number of nodes in q’s active view is greater than 1**, the node will select a random node from its active view, different from the one he received this shuffle message from, and simply forwards the Shuffle request. **Otherwise, node q accepts the Shuffle** request and send back

Right now is only forwarded when expired

Reply:
I think you are correct. Changed accordingly in https://github.com/n0-computer/iroh/pull/1149/commits/f86cde0afca7cefec33c2f7f3cf1886b3d96d8ba